mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMvRxViewModel.kt

```diff
@@ -664,56 +692,66 @@ abstract class BaseMvRxViewModel<S : MvRxState>(
-        this.subscribe(subscriber)
-    } else {
-        this.subscribeWith(
-            MvRxLifecycleAwareObserver(
-                lifecycleOwner,
-                deliveryMode = deliveryMode,
-                lastDeliveredValue = if (deliveryMode is UniqueOnly) {
-                    if (activeSubscriptions.contains(deliveryMode.subscriptionId)) {
-                        throw IllegalStateException(
-                            "Subscribing with a duplicate subscription id: ${deliveryMode.subscriptionId}. " +
-                                "If you have multiple uniqueOnly subscriptions in a MvRx view that listen to the same properties " +
-                                "you must use a custom subscription id. If you are using a custom MvRxView, make sure you are using the proper" +
-                                "lifecycle owner. See BaseMvRxFragment for an example."
-                        )
-                    }
-                    activeSubscriptions.add(deliveryMode.subscriptionId)
-                    lastDeliveredStates[deliveryMode.subscriptionId] as? T
-                } else {
-                    null
-                },
-                onNext = Consumer { value ->
-                    if (deliveryMode is UniqueOnly) {
-                        lastDeliveredStates[deliveryMode.subscriptionId] = value
-                    }
-                    subscriber(value)
-                },
-                onDispose = {
-                    if (deliveryMode is UniqueOnly) {
-                        activeSubscriptions.remove(deliveryMode.subscriptionId)
-                    }
-                }
-            )
-        )
+    ): Job {
+        val flow = if (lifecycleOwner == null || FORCE_DISABLE_LIFECYCLE_AWARE_OBSERVER) {
+            this
+        } else if (deliveryMode is UniqueOnly) {
+            assertOneActiveSubscription(lifecycleOwner, deliveryMode)
+            val lastDeliveredValue: T? = lastDeliveredValue(deliveryMode)
+            this
+                .dropWhile { it == lastDeliveredValue }
+                .flowWhenStarted(lifecycleOwner)
+                .distinctUntilChanged()
+                .onEach { lastDeliveredStates[deliveryMode.subscriptionId] = it }
+        } else {
+            flowWhenStarted(lifecycleOwner)
+        }
+        return flow
+            .onEach { subscriber(it) }
+            .launchIn(viewModelScope)
     }
 
+    private fun <T> Flow<T>.assertOneActiveSubscription(owner: LifecycleOwner, deliveryMode: UniqueOnly) {
+        owner.lifecycle.addObserver(object : LifecycleObserver {
+            @OnLifecycleEvent(Lifecycle.Event.ON_START)
+            fun onStart() {
+                if (activeSubscriptions.contains(deliveryMode.subscriptionId)) error(duplicateSubscriptionMessage(deliveryMode))
+                activeSubscriptions += deliveryMode.subscriptionId
+            }
+
+            @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+            fun onStop() {
```

Comment:
This seems a bit different from the old behavior, none seem like a huge deal, but want to document them:

The code used to remove the subscription in dispose - this handled the case the the observable was terminated manually. You could in theory then subscribe again with the same subscription with no issue. We might be able to get the same behavior with `invokeOnCompletion` and checking for the `cancel` flag

The old code would track on subscription immediately, and would clear in destroy. The new code leaves a period where `Lifecycle < on_start`, where you can subscribe with the same id multiple times, but won't receive the error until the lifecycle has been triggered. Maybe it would make more sense to use `ON_CREATE` and `ON_DESTROY` lifecycle events?